File: frontend/src/types/measurement.ts
// measurement results types - mirror app/schemas/measurement.py results DTOs

import type { InspectionMethod } from "./enums";

export type MeasurementStatus =
  | "QUEUED"
  | "FIRST_FRAME"
  | "AWAITING_CONFIRM"
  | "PROCESSING"
  | "DONE"
  | "ERROR";

// one row of the airport measurements list - status + mission/inspection context + rollup
export interface MeasurementListItem {
  id: string;
  inspection_id: string;
  mission_id: string;
  mission_name: string;
  inspection_method: InspectionMethod;
  inspection_sequence_order: number;
  status: MeasurementStatus;
  // operator-supplied run name; blank falls back to the inspection label
  label: string | null;
  created_at: string | null;
  has_results: boolean;
  pass_count: number;
  fail_count: number;
  error_message: string | null;
}

export interface ReferencePoint {
  light_name: string;
  latitude: number;
  longitude: number;
  elevation: number;
  lha_id: string | null;
  unit_designator: string | null;
  setting_angle: number | null;
  tolerance: number | null;
}

// one PAPI light's box on the first frame, in percentage coordinates (center + size)
export interface LightBox {
  light_name: string;
  x: number;
  y: number;
  size: number;
}

// lightweight progress poll - status doubles as the phase
export interface MeasurementStatusResponse {
  id: string;
  status: MeasurementStatus;
  error_message: string | null;
}

// first-frame image (presigned GET) + the detected/pre-placed boxes to confirm
export interface MeasurementPreview {
  id: string;
  status: MeasurementStatus;
  first_frame_url: string | null;
  boxes: LightBox[];
}

// the fields the start/confirm flow reads off the full aggregate response
export interface Measurement {
  id: string;
  inspection_id: string;
  status: MeasurementStatus;
  label: string | null;
  error_message: string | null;
}

export interface LightSummary {
  light_name: string;
  setting_angle: number | null;
  tolerance: number | null;
  measured_transition_angle: number | null;
  passed: boolean | null;
}

export interface LightSeriesPoint {
  frame_number: number;
  timestamp: number;
  status: string | null;
  angle: number | null;
  horizontal_angle: number | null;
  intensity: number | null;
  area_pixels: number | null;
  // chromaticity derived from the per-frame rgb triple (normalized r/g)
  chromaticity_x: number | null;
  chromaticity_y: number | null;
  // raw per-frame rgb (0-255) the chromaticity is derived from
  red: number | null;
  green: number | null;
  blue: number | null;
  // ground distance from the drone to the light, meters
  distance_ground: number | null;
}

export interface LightSeries {
  light_name: string;
  setting_angle: number | null;
  tolerance: number | null;
  transition_angle_min: number | null;
  transition_angle_middle: number | null;
  transition_angle_max: number | null;
  passed: boolean | null;
  points: LightSeriesPoint[];
}

export interface DronePathPoint {
  frame_number: number;
  timestamp: number;
  latitude: number;
  longitude: number;
  elevation: number | null;
}

// mission-scale protocol aggregation - mirror app/schemas/mission_results.py
export type DeviceEvaluation = "PASS" | "FAIL" | "PENDING" | "NOT_MEASURED";

export interface MissionResultsHeader {
  airport_icao: string;
  airport_name: string;
  mission_name: string;
  measurement_date: string | null;
  drone_model: string | null;
  optical_sensor: string | null;
  reference_system: string | null;
  certificate_number: string | null;
}

export interface MissionWeatherPlaceholder {
  temperature_c: number | null;
  wind: string | null;
  visibility: string | null;
  conditions: string | null;
}

export interface MissionLightResult {
  lha_id: string | null;
  unit_designator: string | null;
  light_name: string;
  setting_angle: number | null;
  tolerance: number | null;
  measured_transition_angle: number | null;
  transition_angle_min: number | null;
  transition_angle_middle: number | null;
  transition_angle_max: number | null;
  passed: boolean | null;
  not_measured: boolean;
}

export interface MissionGlideSlopeResult {
  measured_glide_slope_angle: number | null;
  configured_glide_slope_angle: number | null;
  glide_slope_angle_tolerance: number | null;
  within_tolerance: boolean | null;
}

export interface DeviceResults {
  agl_id: string | null;
  device_type: string;
  device_label: string;
  inspection_id: string | null;
  inspection_method: InspectionMethod | null;
  measurement_id: string | null;
  status: string;
  evaluation: DeviceEvaluation;
  glide_slope: MissionGlideSlopeResult | null;
  lights: MissionLightResult[];
  placeholder_rows: string[];
}

export interface RunwayResults {
  surface_id: string | null;
  runway_identifier: string | null;
  runway_heading: number | null;
  devices: DeviceResults[];
}

export interface DeviceEvaluationRow {
  device_label: string;
  result: DeviceEvaluation;
  restrictions: string | null;
  recommendations: string | null;
}

export interface MissionResults {
  mission_id: string;
  mission_name: string;
  header: MissionResultsHeader;
  weather: MissionWeatherPlaceholder;
  runways: RunwayResults[];
  evaluation: DeviceEvaluationRow[];
  recommendations: string | null;
}

export interface MeasurementResults {
  id: string;
  inspection_id: string;
  status: MeasurementStatus;
  has_results: boolean;
  // run name + inspection context for the results-header blank-label fallback
  label: string | null;
  inspection_method: InspectionMethod | null;
  inspection_sequence_order: number | null;
  runway_heading: number | null;
  // measured glidepath (mid of PAPI_B max / PAPI_C min) vs the snapshotted
  // configured glide slope ± tolerance, plus the within-tolerance verdict
  measured_glide_slope_angle: number | null;
  configured_glide_slope_angle: number | null;
  glide_slope_angle_tolerance: number | null;
  glide_slope_within_tolerance: boolean | null;
  reference_points: ReferencePoint[];
  summaries: LightSummary[];
  lights: LightSeries[];
  drone_path: DronePathPoint[];
  video_urls: Record<string, string>;
}

File: backend/app/schemas/measurement.py
"""measurement DTOs - create/status/preview/confirm shapes for the video pipeline."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

MeasurementStatusStr = Literal[
    "QUEUED", "FIRST_FRAME", "AWAITING_CONFIRM", "PROCESSING", "DONE", "ERROR"
]


class ReferencePointResponse(BaseModel):
    """one PAPI light's snapshotted ground truth."""

    light_name: str
    latitude: float
    longitude: float
    elevation: float
    lha_id: UUID | None = None
    unit_designator: str | None = None
    setting_angle: float | None = None
    tolerance: float | None = None


class LightBox(BaseModel):
    """a PAPI light's box on the first frame, in percentage coordinates."""

    light_name: str
    x: float
    y: float
    size: float


class LightSummaryResponse(BaseModel):
    """per-light PASS/FAIL rollup vs the snapshotted setting angle +/- tolerance."""

    light_name: str
    setting_angle: float | None = None
    tolerance: float | None = None
    measured_transition_angle: float | None = None
    passed: bool | None = None


class MeasurementResponse(BaseModel):
    """full measurement aggregate as returned to the operator."""

    id: UUID
    inspection_id: UUID
    status: MeasurementStatusStr
    runway_heading: float | None = None
    reference_points: list[ReferencePointResponse] = []
    light_boxes: list[LightBox] = []
    summaries: list[LightSummaryResponse] = []
    object_key: str | None = None
    first_frame_object_key: str | None = None
    error_message: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class MeasurementStatusResponse(BaseModel):
    """lightweight progress poll - status doubles as the phase."""

    id: UUID
    status: MeasurementStatusStr
    error_message: str | None = None


class MeasurementListItemResponse(BaseModel):
    """one row of the airport measurements list - status + mission/inspection context + rollup.

    ``has_results`` and the PASS/FAIL counts derive from the aggregate's summaries +
    object_key, so the list page can route each row without a per-row results fetch.
    """

    id: UUID
    inspection_id: UUID
    mission_id: UUID
    mission_name: str
    inspection_method: str
    inspection_sequence_order: int
    status: MeasurementStatusStr
    created_at: datetime | None = None
    has_results: bool = False
    pass_count: int = 0
    fail_count: int = 0
    error_message: str | None = None


class ConfirmLightsRequest(BaseModel):
    """operator-confirmed/adjusted first-frame light boxes that start full processing."""

    boxes: list[LightBox]


class MeasurementPreviewResponse(BaseModel):
    """first-frame image (presigned GET) plus the detected/pre-placed boxes."""

    id: UUID
    status: MeasurementStatusStr
    first_frame_url: str | None = None
    boxes: list[LightBox] = []


# results shapes - the pivoted view of the gzipped per-frame blob in object storage


class LightSeriesPoint(BaseModel):
    """one frame's reading for a single light - the unit the timeseries plots.

    chromaticity is derived from the per-frame rgb triple (normalized r/g), there
    is no direct chromaticity key in the engine blob.
    """

    frame_number: int
    timestamp: float
    status: str | None = None
    angle: float | None = None
    horizontal_angle: float | None = None
    intensity: float | None = None
    area_pixels: int | None = None
    chromaticity_x: float | None = None
    chromaticity_y: float | None = None


class LightSeries(BaseModel):
    """one light's full timeseries plus its transition angles and PASS/FAIL rollup."""

    light_name: str
    setting_angle: float | None = None
    tolerance: float | None = None
    transition_angle_min: float | None = None
    transition_angle_middle: float | None = None
    transition_angle_max: float | None = None
    passed: bool | None = None
    points: list[LightSeriesPoint] = []


class DronePathPoint(BaseModel):
    """one frame's drone position, the ordered flown path on the map."""

    frame_number: int
    timestamp: float
    latitude: float
    longitude: float
    elevation: float | None = None


class MeasurementResultsResponse(BaseModel):
    """full results payload for the operator results page.

    ``has_results`` is false (and ``lights`` / ``drone_path`` / ``video_urls`` empty)
    until the run is DONE with a results blob in object storage.
    """

    id: UUID
    inspection_id: UUID
    status: MeasurementStatusStr
    has_results: bool = False
    runway_heading: float | None = None
    reference_points: list[ReferencePointResponse] = []
    summaries: list[LightSummaryResponse] = []
    lights: list[LightSeries] = []
    drone_path: list[DronePathPoint] = []
    video_urls: dict[str, str] = {}
